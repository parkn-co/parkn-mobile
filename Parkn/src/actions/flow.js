export type Action = {
  type: String,
  payload: ?Any
};

export type AsyncAction = (dispatch: Function, getState: Function) => void
