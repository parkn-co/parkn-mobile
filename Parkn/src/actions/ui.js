// @flow
import type {Action} from 'flow-declarations/redux';

export const SWITCH_TO_VIEW = 'SWITCH_TO_VIEW';
export const switchToView = (payload: Object): Action => ({
  type: SWITCH_TO_VIEW,
  payload,
});
