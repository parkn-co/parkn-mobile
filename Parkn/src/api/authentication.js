import Api from './base';

export function signIn({email, password}) {
  return Api.post('auth/signin', {email, password})
    .then(res => res);
}

export function signUp({email, firstName, lastName, password}) {
  return Api.post('auth/signup', {email, firstName, lastName, password})
    .then(res => res);
}

export function signOut(getState) {
  return Api.authPost(getState, 'auth/signout');
}

export function fetchUser(getState) {
  return Api.authGet(getState, 'users')
    .then(res => res);
}

export function storeToken(token) {
  Api.storage.set('token', token);
}

export function fetchToken() {
  return Api.storage.get('token');
}
