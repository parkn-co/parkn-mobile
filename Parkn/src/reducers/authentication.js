// Import action handlers
import {
  AUTH_SET_FORM_VALUES,
  AUTH_SIGN_IN,
  AUTH_SET_ERRORS,
  AUTH_FETCH_USER,
  AUTH_IS_FETCHING,
} from '../actions/authentication';

import {REHYDRATE} from 'redux-persist/constants';

const initialState = {
  form: {
    firstName: '',
    lastName: '',
    email: '',
    password: '',
    confirmPassword: '',
  },
  formType: '',
  errors: {},
  token: null,
  user: {},
  isFetching: false,
};

// Action Handlers
const REDUCER_ACTION_HANDLERS = {
   [AUTH_SET_FORM_VALUES]: (state, {payload}) => ({
     ...state,
     form: {...state.form, ...payload},
     isFetching: false,
   }),
   [AUTH_SIGN_IN]: (state, {payload: {token}}) => ({
     ...state,
     token,
     isFetching: false,
   }),
   [AUTH_SET_ERRORS]: (state, {payload}) => ({
     ...state,
     errors: payload,
     isFetching: false,
   }),
   [AUTH_FETCH_USER]: (state, {payload}) => ({
     ...state,
     user: payload,
     isFetching: false,
   }),
   [AUTH_IS_FETCHING]: state => ({
     ...state,
     isFetching: true,
   }),
   [REHYDRATE]: (state, {payload}) => ({
     ...initialState,
     token: payload && payload.authentication && payload.authentication.token,
   }),
};

// Reducer
export default function authenticationReducer(state = initialState, action) {
  const handler = REDUCER_ACTION_HANDLERS[action.type]
  return handler ? handler(state, action) : state
}
