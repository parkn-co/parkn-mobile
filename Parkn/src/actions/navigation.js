// @flow
import {Action} from './flow';

export const NAV_NAVIGATE_TO = 'NAVIGATE_TO';
export function navigateTo(route: Object): Action {
  return {
    type: NAV_NAVIGATE_TO,
    payload: route,
  };
}

export const NAV_DID_NAVIGATE_TO = 'DID_NAVIGATE_TO';
export function didNavigateTo(route: Object): Action {
  return {
    type: NAV_DID_NAVIGATE_TO,
    payload: route,
  }
}

export const NAV_SET_IS_NAVIGATING = 'NAV_SET_IS_NAVIGATING';
export function setIsNavigating(isNavigating: Boolean): Action {
  return {
    type: NAV_SET_IS_NAVIGATING,
    payload: isNavigating,
  }
}
