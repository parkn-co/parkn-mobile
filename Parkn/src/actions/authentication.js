import * as AuthApi from '../api/authentication';

export const AUTH_IS_FETCHING = 'AUTH_IS_FETCHING';
export function setIsFetching() {
  return {
    type: AUTH_IS_FETCHING,
  };
}

export const AUTH_SET_FORM_VALUES = 'AUTH_SET_FORM_VALUES';
export function setFormValues(values) {
  return {
    type: AUTH_SET_FORM_VALUES,
    payload: values,
  };
}

export const AUTH_SIGN_IN = 'AUTH_SIGN_IN';
export const AUTH_SET_ERRORS = 'AUTH_SET_ERRORS';
export function authenticateWithValues(isSignUp, values) {
  return (dispatch, getState) => {
    dispatch({
      type: AUTH_SET_FORM_VALUES,
      payload: values
    });

    dispatch(setIsFetching());

    const {
      authentication: {form}
    } = getState();

    let method = 'signIn';
    if (isSignUp) {
      method = 'signUp';
    }

    AuthApi[method]({...form, ...values})
    .then(token => dispatch({
      type: AUTH_SIGN_IN,
      payload: token,
    }))
    .catch(err => {
      let payload = err;

      if (err.status === 409) {
        payload = {...err, email: err.error};
      } else if (err.status === 404 || err.status === 500) {
        payload = {...err, email: 'Incorrect email or password'};
      }

      dispatch({
        type: AUTH_SET_ERRORS,
        payload,
      })
    });
  }
}

export const AUTH_FETCH_USER = 'AUTH_FETCH_USER';
export function fetchUser() {
  return (dispatch, getState) => {
    dispatch(setIsFetching());

    AuthApi.fetchUser(getState)
    .then(token => dispatch({
      type: AUTH_FETCH_USER,
      payload: token,
    }))
    .catch(err => dispatch({
      type: AUTH_SET_ERRORS,
      payload: err,
    }));
  }
}
