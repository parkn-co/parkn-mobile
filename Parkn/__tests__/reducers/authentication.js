/* eslint-env jest */

import 'react-native';
import React from 'react';
import { REHYDRATE } from 'redux-persist/constants';
import * as authenticationActions from '../../src/actions/authentication';
import authenticationReducer from '../../src/reducers/authentication';

describe('Authentication reducer', () => {
  const blankState = {
    form: {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      confirmPassword: '',
    },
    errors: {},
    token: null,
    user: {},
    isFetching: false,
  };

  it('returns the correct new state when AUTH_SET_FORM_VALUES is dispatched', () => {
    const action = {
      type: authenticationActions.AUTH_SET_FORM_VALUES,
      payload: {
        firstName: 'jared',
        lastName: 'ramirez',
        email: 'jared@example.com',
        password: 'foo',
        confirmPassword: 'foo',
      },
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when AUTH_SIGN_IN is dispatched', () => {
    const action = {
      type: authenticationActions.AUTH_SIGN_IN,
      payload: {
        token: 'WOOT_THIS_TOKEN_GIVES_YOU_ALLLLLLL_ACCESS_;)',
      },
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when AUTH_SIGN_OUT is dispatched', () => {
    const action = {
      type: authenticationActions.AUTH_SIGN_OUT,
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when AUTH_SET_ERRORS is dispatched', () => {
    const action = {
      type: authenticationActions.AUTH_SET_ERRORS,
      payload: {
        firstName: 'Homie, firstname can\'t be more than 1000 characters',
        lastName: 'You forgot to enter you last name!!!!',
      },
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when AUTH_FETCH_USER is dispatched', () => {
    const action = {
      type: authenticationActions.AUTH_FETCH_USER,
      payload: {
        id: 1029581035,
        firstName: 'jared',
        lastName: 'ramirez',
      },
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when AUTH_IS_FETCHING is dispatched', () => {
    const action = {
      type: authenticationActions.AUTH_IS_FETCHING,
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });

  it('returns the correct new state when REHYDRATE is dispatched', () => {
    const action = {
      type: REHYDRATE,
      payload: {
        authentication: {
          token: 'this_is_a_rad_token',
        },
      },
    };
    expect(authenticationReducer(blankState, action)).toMatchSnapshot();
  });
});
