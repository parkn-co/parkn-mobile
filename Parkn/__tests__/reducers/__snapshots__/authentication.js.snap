exports[`Authentication reducer returns the correct new state when AUTH_FETCH_USER is dispatched 1`] = `
Object {
  "errors": Object {},
  "form": Object {
    "confirmPassword": "",
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "isFetching": false,
  "token": null,
  "user": Object {
    "firstName": "jared",
    "id": 1029581035,
    "lastName": "ramirez",
  },
}
`;

exports[`Authentication reducer returns the correct new state when AUTH_IS_FETCHING is dispatched 1`] = `
Object {
  "errors": Object {},
  "form": Object {
    "confirmPassword": "",
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "isFetching": true,
  "token": null,
  "user": Object {},
}
`;

exports[`Authentication reducer returns the correct new state when AUTH_SET_ERRORS is dispatched 1`] = `
Object {
  "errors": Object {
    "firstName": "Homie, firstname can\'t be more than 1000 characters",
    "lastName": "You forgot to enter you last name!!!!",
  },
  "form": Object {
    "confirmPassword": "",
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "isFetching": false,
  "token": null,
  "user": Object {},
}
`;

exports[`Authentication reducer returns the correct new state when AUTH_SET_FORM_VALUES is dispatched 1`] = `
Object {
  "errors": Object {},
  "form": Object {
    "confirmPassword": "foo",
    "email": "jared@example.com",
    "firstName": "jared",
    "lastName": "ramirez",
    "password": "foo",
  },
  "isFetching": false,
  "token": null,
  "user": Object {},
}
`;

exports[`Authentication reducer returns the correct new state when AUTH_SIGN_IN is dispatched 1`] = `
Object {
  "errors": Object {},
  "form": Object {
    "confirmPassword": "",
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "isFetching": false,
  "token": "WOOT_THIS_TOKEN_GIVES_YOU_ALLLLLLL_ACCESS_;)",
  "user": Object {},
}
`;

exports[`Authentication reducer returns the correct new state when AUTH_SIGN_OUT is dispatched 1`] = `
Object {
  "errors": Object {},
  "form": Object {
    "confirmPassword": "",
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "isFetching": false,
  "token": null,
  "user": Object {},
}
`;

exports[`Authentication reducer returns the correct new state when REHYDRATE is dispatched 1`] = `
Object {
  "errors": Object {},
  "form": Object {
    "confirmPassword": "",
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "isFetching": false,
  "token": "this_is_a_rad_token",
  "user": Object {},
}
`;
